File: Hogweed.cpp
#include "CONSTANTS.h"
#include "Hogweed.h"
#include "World.h"
#include "Animal.h"

#include <vector>

Hogweed::Hogweed(World &world, const Point &position) : Plant(world, position, HOGWEED_SYMBOL, HOGWEED_STRENGTH, HOGWEED_COLOR, HOGWEED_NAME) {}

void Hogweed::Action() {
  if(CanAction()) {
    std::vector<Point> newPositions = this->world.GetPositionsAround(this->GetPosition());
    for (int i = 0; i < 4; i++) {
      Point newPosition = newPositions[i];
      Organism *organism = this->world.GetOrganismAt(newPosition);
      // Check if there is an animal on the field and it is not a cyber-sheep
      if(organism != nullptr && dynamic_cast<Animal*>(organism) != nullptr && organism->GetSymbol() != CYBER_SHEEP_SYMBOL) {
        organism->Die();
        this->world.messageManager.AddDeathMessage(organism->GetName(), this->GetName());
      }
    }
  }
}

bool Hogweed::Collision(Organism *attackerOrganism) {
  this->world.messageManager.AddPlantEatingMessage(this->GetName(), attackerOrganism->GetName());
  this->Die();
  if(attackerOrganism->GetSymbol() != CYBER_SHEEP_SYMBOL) {
    attackerOrganism->Collision(this);
  }
  return false;
}

Organism* Hogweed::Clone(const Point &position) {
  return new Hogweed(this->world, position);
}

File: Java/Hogweed.java
import java.util.Vector;
public class Hogweed extends Plant {
  public Hogweed(World world, Point position) {
    super(world, position, Constants.HOGWEED_SYMBOL, Constants.HOGWEED_STRENGTH, Constants.HOGWEED_COLOR, Constants.HOGWEED_NAME);
  }

  @Override
  public Organism Clone(Point position) {
    return new Hogweed(this.world, position);
  }

  @Override
  public boolean Collision(Organism other) {
    this.world.messageManager.AddPlantEatingMessage(other.GetName(), this.GetName());
    this.world.messageManager.AddDeathMessage(other.GetName(), this.GetName());
    this.Die();
    if(other.GetSymbol() != Constants.CYBERSHEEP_SYMBOL) {
      other.Die();
    }
    return false;
  }

  @Override
  public void Action() {
    if(CanAction()) {
      Vector<Point> positions = world.GetPositionsAround(this.GetPosition());
      for(Point position : positions) {
        Organism other = world.GetOrganismAt(position);
        // Check if there is an animal on the field, and it is not a cyber-sheep
        if(other instanceof Animal && other.GetSymbol() != Constants.CYBERSHEEP_SYMBOL) {
          other.Die();
          this.world.messageManager.AddDeathMessage(other.GetName(), this.GetName());
        }
      }

    }
  }
}
